src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -153,8 +154,71 @@ class FastSync(
       assignedHandlers -= handler
     }
 
+    private def insertHeader(header: BlockHeader): Boolean = {
+      val parentTdOpt = blockchain.getTotalDifficultyByHash(header.parentHash)
+      parentTdOpt foreach { parentTotalDifficulty =>
+        blockchain.save(header)
+        blockchain.save(header.hash, parentTotalDifficulty + header.difficulty)
+
+        if (header.number > syncState.bestBlockHeaderNumber) {
+          syncState = syncState.copy(bestBlockHeaderNumber = header.number)
+        }
+
+        syncState = syncState
+          .enqueueBlockBodies(Seq(header.hash))
+          .enqueueReceipts(Seq(header.hash))
+      }
+
+      parentTdOpt.isDefined
+    }
+
+    @tailrec
+    private def processHeaders(peer: Peer, headers: Seq[BlockHeader]): Unit = {
+      if (headers.nonEmpty) {
+        val header = headers.head
+        val remaining = headers.tail
+
+        val shouldValidate = header.number >= syncState.nextBlockToFullyValidate
+        if (shouldValidate) {
+          syncState = syncState.copy(
+            nextBlockToFullyValidate =
+              if (syncState.bestBlockHeaderNumber >= syncState.targetBlock.number - X) header.number + 1
+              else header.number + K / 2 + Random.nextInt(K))
+
+          validators.blockHeaderValidator.validate(header, blockchain) match {
+            case Right(_) =>
+              if (insertHeader(header)) processHeaders(peer, remaining)
+              else ()
+
+            case Left(error) =>
+              log.warning(s"Block header validation failed during fast sync at block ${header.number}: $error")
+              blacklist(peer.id, blacklistDuration, "block header validation failed")
+
+              // discard last N blocks
+              (header.number to ((header.number - N) max 1) by -1).foreach { n =>
+                blockchain.getBlockHeaderByNumber(n).foreach { header =>
```

Comment:
Is this `getBlockHeaderByNumber` needed if we already have `header.hash` and `removeBlock` doesn't fail (afaik) if hash does not exist?

Reply:
good catch, it's not needed. (I don't expect this piece of code to be called too often, but still)

Reply:
Actually it is needed here. `header` was shadowed inside `foreach`, we only have one header (the one we start with). I changed to name to make that more clear